File: client/entities/FlightServant.cpp
#include "FlightServant.hpp"

#include "Flight.hpp"
#include <string>
#include <iostream>
#include <iomanip>

FlightServant::FlightServant()
{
    this->flightId = "";
    this->source = "";
    this->destination = "";
    this->seatsAvailable = -1;
    this->seatsBooked = -1;
    this->price = -1;
    this->departureTime = Time(0);
    this->duration = Time(0);
}
FlightServant::FlightServant(char* flightId, char* source, char* destination, int seatsAvailable, int seatsBooked, float price, int departureTime, int duration)
{
    this->flightId = flightId;
    this->source = source;
    this->destination = destination;
    this->seatsAvailable = seatsAvailable;
    this->seatsBooked = seatsBooked;
    this->price = price;
    this->departureTime = Time(departureTime);
    this->duration = Time(duration);
}
char* FlightServant::getFlightID()
{
    return this->flightId;
}
void FlightServant::setFlightID(char* flightId)
{
    strcpy(this->flightId, flightId);
}
char* FlightServant::getSource()
{
    return this->source;
}
void FlightServant::setSource(char* source)
{
    strcpy(this->source, source);
}
char* FlightServant::getDestination()
{
    return this->destination;
}
void FlightServant::setDestination(char* destination)
{
    strcpy(this->destination, destination);
}
int FlightServant::getSeatsAvailable()
{
    return this->seatsAvailable;
}
void FlightServant::setSeatsAvailable(int seatsAvaiable)
{
    this->seatsAvailable = seatsAvaiable;
}
int FlightServant::getSeatsBooked()
{
    return this->seatsBooked;
}
void FlightServant::setSeatsBooked(int seatsBooked)
{
    this->seatsBooked = seatsBooked;
}
float FlightServant::getPrice()
{
    return this->price;
}
void FlightServant::setPrice(float price)
{
    this->price = price;
}
Time FlightServant::getDepartureTime()
{
    return this->departureTime;
}
void FlightServant::setDepartureTime(Time time)
{
    this->departureTime = time;
}
Time FlightServant::getDuration()
{
    return this->duration;
}
void FlightServant::setDuration(Time time)
{
    this->duration = time;
}

void FlightServant::display()
{
    std::cout << std::left << std::setw(19) << "Flight Number" << ':' << '\t' << this->flightId << '\n';
    std::cout << std::left << std::setw(19) << "Source" << ':' << '\t' << this->source << '\n';
    std::cout << std::left << std::setw(19) << "Destination" << ':' << '\t' << this->destination << '\n';
    std::cout << std::left << std::setw(19) << "Departure Time" << ':' << '\t' << this->departureTime.to_string("%H:%M:%S") << '\n';
    std::cout << std::left << std::setw(19) << "Duration" << ':' << '\t' << this->duration.to_string("%H:%M hrs") << '\n';
    std::cout << std::left << std::setw(19) << "Available Seats" << ':' << '\t' << this->seatsAvailable << '\n';
    std::cout << std::left << std::setw(19) << "Booked Seats" << ':' << '\t' << this->seatsBooked << '\n';
    std::cout << std::left << std::setw(19) << "Price" << ':' << '\t' << this->price << "$" << '\n';

    // std::cout << this->flightId << "@" << this->source << "->" << this->destination << std::endl;
    // std::cout << this->seatsAvailable << "-" << this->seatsBooked << "#" << this->price << std::endl;
}

size_t FlightServant::serialization_size() const
{
    return SerializePOD<char* >::serialization_size(flightId) +
           SerializePOD<char* >::serialization_size(source) +
           SerializePOD<char* >::serialization_size(destination) +
           SerializePOD<int>::serialization_size(seatsAvailable) +
           SerializePOD<int>::serialization_size(seatsBooked) +
           SerializePOD<float>::serialization_size(price);
}
char* FlightServant::serialize() const
{
    int size = serialization_size();

    // Initialize output buffer for serialized string
    char* dataOut = new char[size + 1];
    dataOut[size] = '\0';

    SerializePOD<char* >::serialize(dataOut, flightId);
    SerializePOD<char* >::serialize(dataOut, source);
    SerializePOD<char* >::serialize(dataOut, destination);
    SerializePOD<int>::serialize(dataOut, seatsAvailable);
    SerializePOD<int>::serialize(dataOut, seatsBooked);
    SerializePOD<float>::serialize(dataOut, price);
    SerializePOD<int>::serialize(dataOut, departureTime.toSecondOfDay());
    SerializePOD<int>::serialize(dataOut, duration.toSecondOfDay());

    // Reset pointer to start of serialized string and return
    dataOut -= size;
    return dataOut;
}

void FlightServant::deserialize(char* &dataIn)
{
    char* flightId;
    char* source;
    char* destination;
    int seatsAvailable;
    int seatsBooked;
    float price;
    int int_departureTime;
    int int_duration;

    SerializePOD<char*>::deserialize(dataIn, flightId);
    SerializePOD<char*>::deserialize(dataIn, source);
    SerializePOD<char*>::deserialize(dataIn, destination);
    SerializePOD<int>::deserialize(dataIn, seatsAvailable);
    SerializePOD<int>::deserialize(dataIn, seatsBooked);
    SerializePOD<float>::deserialize(dataIn, price);
    SerializePOD<int>::deserialize(dataIn, int_departureTime);
    SerializePOD<int>::deserialize(dataIn, int_duration);

    this->flightId = flightId;
    this->source = source;
    this->destination = destination;
    this->seatsAvailable = seatsAvailable;
    this->seatsBooked = seatsBooked;
    this->price = price;
    this->departureTime = Time(int_departureTime);
    this->duration = Time(int_duration);
}

File: server/entities/FlightServant.java
package entities;
import java.io.IOException;
import java.time.LocalTime;
import marshall.SerializePOD;
import utils.PrimitiveSizes;

/* 
    Servant class for Flight interface.
    Defines state variables for class including flight ID, source, destination, number of seats currently available,
    number of seats booked, price of seat, departure time, and duration.
    Implements methods including getters and setters, display, and marshalling-unmarshalling.
*/
public class FlightServant implements Flight {
    private char[] flightId;
    private char[] source;
    private char[] destination;
    private int seatsAvailable;
    private int seatsBooked;
    private float price;
    private LocalTime departureTime;
    private LocalTime duration;

    @Override
    public long size() {
        return (
            PrimitiveSizes.sizeof((long)flightId.length) + flightId.length + 
            PrimitiveSizes.sizeof((long)source.length) + source.length +
            PrimitiveSizes.sizeof((long)destination.length) + destination.length +
            PrimitiveSizes.sizeof(seatsAvailable) + 
            PrimitiveSizes.sizeof(seatsBooked) + 
            PrimitiveSizes.sizeof(price) +
            PrimitiveSizes.sizeof(departureTime.toSecondOfDay()) +
            PrimitiveSizes.sizeof(duration.toSecondOfDay())
        );
    }

    public FlightServant()
    {
        
    }

    public FlightServant(char[] flightId, char[] source, char[] destination, int seatsAvailable, int seatsBooked, float price, LocalTime departureTime, LocalTime duration) {
        this.flightId = flightId;
        this.source = source;
        this.destination = destination;
        this.seatsAvailable = seatsAvailable;
        this.seatsBooked = seatsBooked;
        this.price = price;
        this.duration = duration;
        this.departureTime = departureTime;
    }

    public char[] getFlightId() {
        return flightId;
    }

    public void setFlightId(char[] flightId) {
        this.flightId = flightId;
    }

    public char[] getSource() {
        return source;
    }

    public void setSource(char[] source) {
        this.source = source;
    }

    public char[] getDestination() {
        return destination;
    }

    public void setDestination(char[] destination) {
        this.destination = destination;
    }

    public int getSeatsAvailable() {
        return seatsAvailable;
    }

    public void setSeatsAvailable(int seatsAvailable) {
        this.seatsAvailable = seatsAvailable;
    }

    public int getSeatsBooked() {
        return seatsBooked;
    }

    public void setSeatsBooked(int seatsBooked) {
        this.seatsBooked = seatsBooked;
    }

    public float getPrice() {
        return price;
    }

    public void setPrice(float price) {
        this.price = price;
    }

    public LocalTime getDepartureTime() {
        return this.departureTime;
    }
    public void setDepartureTime(LocalTime departureTime) {
        this.departureTime = departureTime;
    }

    public LocalTime getDuration() {
        return this.duration;
    }

    public void setDuration(LocalTime duration) {
        this.duration = duration;
    }

    @Override
    public void display() {
        System.out.print(new String(flightId) + " @ " + new String(source) + " --> " + new String(destination) + "\n");
        System.out.print(seatsAvailable + " - " + seatsBooked + " # " + price + "\n");
        System.out.println(departureTime + " " + duration);
    }

    @Override
    public void deserialize(byte[] dataIn) {
        int start = 0;
        
        flightId = SerializePOD.deserializeString(dataIn, start);
        long flightIdSize = flightId.length;
        start += flightIdSize + PrimitiveSizes.sizeof(flightIdSize);

        source = SerializePOD.deserializeString(dataIn, start);
        long sourceSize = source.length;
        start += sourceSize + PrimitiveSizes.sizeof(sourceSize);

        destination = SerializePOD.deserializeString(dataIn, start);
        long destinationSize = destination.length;
        start += destinationSize + PrimitiveSizes.sizeof(destinationSize);

        seatsAvailable = SerializePOD.deserializeInt(dataIn, start);
        start += PrimitiveSizes.sizeof(seatsAvailable);

        seatsBooked = SerializePOD.deserializeInt(dataIn, start);
        start += PrimitiveSizes.sizeof(seatsBooked);

        price = SerializePOD.deserializeFloat(dataIn, start);
        start += PrimitiveSizes.sizeof(price);

        int int_departureTime = SerializePOD.deserializeInt(dataIn, start);
        departureTime = LocalTime.ofSecondOfDay(int_departureTime);
        start += PrimitiveSizes.sizeof(int_departureTime);
        
        int int_duration = SerializePOD.deserializeInt(dataIn, start);
        duration = LocalTime.ofSecondOfDay(int_duration);
        start += PrimitiveSizes.sizeof(int_duration);
    }

    public byte[] serialize() throws IOException {
        byte[] buffer = new byte[(int)this.size()];

        byte[] flightIdBuffer = SerializePOD.serialize(flightId);
        byte[] sourceBuffer = SerializePOD.serialize(source);
        byte[] destinationBuffer = SerializePOD.serialize(destination);
        byte[] seatsAvailableBuffer = SerializePOD.serialize(seatsAvailable);
        byte[] seatsBookedBuffer = SerializePOD.serialize(seatsBooked);
        byte[] priceBuffer = SerializePOD.serialize(price);
        byte[] departureTimeBuffer = SerializePOD.serialize(departureTime.toSecondOfDay());
        byte[] durationBuffer = SerializePOD.serialize(duration.toSecondOfDay());

        int i = 0;
        System.arraycopy(flightIdBuffer, 0, buffer, i, flightIdBuffer.length);
        i += flightIdBuffer.length;
        System.arraycopy(sourceBuffer, 0, buffer, i, sourceBuffer.length);
        i += sourceBuffer.length;
        System.arraycopy(destinationBuffer, 0, buffer, i, destinationBuffer.length);
        i += destinationBuffer.length;
        System.arraycopy(seatsAvailableBuffer, 0, buffer, i, seatsAvailableBuffer.length);
        i += seatsAvailableBuffer.length;
        System.arraycopy(seatsBookedBuffer, 0, buffer, i, seatsBookedBuffer.length);
        i += seatsBookedBuffer.length;
        System.arraycopy(priceBuffer, 0, buffer, i, priceBuffer.length);
        i += priceBuffer.length;
        System.arraycopy(departureTimeBuffer, 0, buffer, i, departureTimeBuffer.length);
        i += departureTimeBuffer.length;
        System.arraycopy(durationBuffer, 0, buffer, i, durationBuffer.length);
        i += durationBuffer.length;

        return buffer;
    }

}

